pilot/adapter/serviceregistry/aggregate/controller.go

```diff
@@ -15,167 +15,527 @@
+)
+
 // Registry specifies the collection of service registry related interfaces
 type Registry struct {
 	Name platform.ServiceRegistry
 	model.Controller
-	model.ServiceDiscovery
-	model.ServiceAccounts
+	// The service mesh view that this registry belongs to
+	MeshView *MeshResourceView
 }
 
-// Controller aggregates data across different registries and monitors for changes
-type Controller struct {
+// MeshResourceView is an aggregated store for resources sourced from various registries
+type MeshResourceView struct {
 	registries []Registry
+
+	// Mutex guards services, serviceInstances and corresponding labels
+	mu sync.RWMutex
+
+	// Canonical map of mesh service key to service references
+	services map[resourceKey]*model.Service
+
+	// Canonical map of mesh service instance keys to service instance references
+	serviceInstances map[resourceKey]*model.ServiceInstance
+
+	// A reverse map that associates label names to label values and their associated service resource keys
+	serviceLabels nameValueKeysMap
+
+	// A reverse map that associates label names to label values and their associated service instance resource keys
+	serviceInstanceLabels nameValueKeysMap
+
+	// Cache eviction handler for notifying changes to model objects
+	cacheEvictionHandler model.CacheEvictionHandler
 }
 
-// NewController creates a new Aggregate controller
-func NewController() *Controller {
-	return &Controller{
-		registries: make([]Registry, 0),
+type serviceList []*model.Service
+
+func (sl *serviceList) String() string {
+	out := "["
+	for _, s := range *sl {
+		out = out + s.Hostname + ", "
 	}
+	out = out + "]"
+	return out
 }
 
-// AddRegistry adds registries into the aggregated controller
-func (c *Controller) AddRegistry(registry Registry) {
-	c.registries = append(c.registries, registry)
+type serviceInstanceList []*model.ServiceInstance
+
+func (sl *serviceInstanceList) String() string {
+	out := "["
+	for _, i := range *sl {
+		out = out + i.Service.Hostname + ":" + i.Endpoint.Address + ":" + fmt.Sprintf("%d", i.Endpoint.Port) + ", "
+	}
+	out = out + "]"
+	return out
 }
 
-// Services lists services from all platforms
-func (c *Controller) Services() ([]*model.Service, error) {
-	services := make([]*model.Service, 0)
-	var errs error
-	for _, r := range c.registries {
-		svcs, err := r.Services()
-		if err != nil {
-			errs = multierror.Append(errs, err)
-		} else {
-			services = append(services, svcs...)
-		}
+// NewMeshResourceView creates an aggregated store for resources sourced from various registries
+func NewMeshResourceView() *MeshResourceView {
+	return &MeshResourceView{
+		registries:            make([]Registry, 0),
+		mu:                    sync.RWMutex{},
+		services:              make(map[resourceKey]*model.Service),
+		serviceInstances:      make(map[resourceKey]*model.ServiceInstance),
+		serviceLabels:         make(nameValueKeysMap),
+		serviceInstanceLabels: make(nameValueKeysMap),
```

Comment:
I don't understand the need for this refactoring.. May be an easier way to get PoC is to add a "name" field to the old Controller struct in this file? Such that you can refer to each controller by its name. I would not pay too much attention to the cache handlers, etc. as they are going away soon.